src/main/java/net/minecraftforge/resource/ResourceCacheManager.java

```diff
@@ -0,0 +1,406 @@
+         */
+        private final String namespace;
+        /**
+         * Indicates if the indexing should run off-thread.
+         */
+        private final boolean indexOffThread;
+        /**
+         * The path builder to use.
+         */
+        private final BiFunction<PackType, String, Path> pathBuilder;
+        /**
+         * The path walker stream factory to use.
+         */
+        private final PathWalkerFactory pathFinder;
+
+        /**
+         * The cache entries for this manager.
+         */
+        private final Map<String, List<ResourceCacheEntry>> entriesByPathPrefix = Maps.newConcurrentMap();
+        /**
+         * Indicates if the cache has been loaded successfully.
+         */
+        private final AtomicBoolean cacheLoaded = new AtomicBoolean(false);
+
+        /**
+         * Creates a new namespaced resource cache manager.
+         *
+         * @param packType       The pack type that this manager handles.
+         * @param namespace      The namespace that this manager handles.
+         * @param indexOffThread True to enable indexing off-thread.
+         * @param pathBuilder    The path builder to use.
+         * @param pathFinder     The path walker stream factory to use.
+         */
+        private NamespacedResourceCacheManager(final PackType packType, final String namespace, final boolean indexOffThread, BiFunction<PackType, String, Path> pathBuilder, final PathWalkerFactory pathFinder)
+        {
+            this.packType = packType;
+            this.namespace = namespace;
+            this.indexOffThread = indexOffThread;
+            this.pathBuilder = pathBuilder;
+            this.pathFinder = pathFinder;
+        }
+
+        /**
+         * Triggers indexing of this manager.
+         */
+        public void index()
+        {
+            if (indexOffThread)
+            {
+                // Run off-thread.
+                CompletableFuture.runAsync(
+                        this::doIndex,
+                        Util.backgroundExecutor()
+                );
+            } else
+            {
+                // Run on-thread
+                doIndex();
+            }
+        }
+
+        /**
+         * Performs the actual indexing of this manager.
+         * Can be invoked on- or off-thread.
+         */
+        private void doIndex()
+        {
+            // Get the path to the root of the namespace in the current pack.
+            final Path rootPath = pathBuilder.apply(packType, namespace);
+
+            // Stream element resource that combines a normalized path and a joined path using the "/" as separator.
+            record PathWithLocationPath(Path path, String locationPath)
+            {
+            }
+
+            // Build a walkable stream, process it
+            try (final Stream<Path> paths = pathFinder.createWalkingStream(rootPath))
+            {
+                paths.parallel() // Run the stream in parallel
```

Comment:
You are correct that can be an ArrayList if I am not mistaken, at least with the current setup, since the resource access is guarded behind a hasCached check, I will adapt this.

Reply:
Well for that to be safe the stream should *not* be parallel, though (that's why the comment was on that line)